
# SHaDiE
## "Simulating Haploid-Diploid Evolution"

### Description of project goal:
`shadie` will be a wrapper around the forward-in-time evolutionary simulation program [SLiM3](https://messerlab.org/slim/). It will accept a phylogeny from the user and generate equivalent population demography inside a script that can be fed directly into SLiM3. THe user will also be able to define a number of different parameters for the simulation and the program will prepare the slim script appropriately along with a dispatch script, which is necessary to run SLiM3 from the command line. The user will be able to run the SLiM simulation using a `shadie` command. After the simulation is complete, `shadie` will also provide a number of summary statistics and methods for inspecting the output of the simulation. 

`shadie` is specifically designed to model the varied biphasic lifecycles in plants, allowing the user to explore the evolutionary consequences of different life histories. 

~

### Description of the code:
**Dependencies:**
* `numpy`: to manipulate data
* `pandas`: to organize and analyze data
* `toytree`: to generate phylogenies
* `tskit`: provides access to the tree-sequence recording from SLiM3 simulation (.treese file)
* `pyslim`: process and read .trees objects output by SLiM3; uses `tskit` and `msprime` under the hood
* `msprime`: used by pyslim under the hood to overlay neutral mutation using coalescent. Can also be used for recapitation
* `toyplot`: used to generate static chromosome plots
* `altair`: used to generate interactive chromosome and distribution plots


**Classes:**
- `Main`: reads in user-provided info from `Demography`, `Chromosome`, and `Reproduction to generate the .slim script file that will be used to run the SLiM3 simulation
- `Postsim`: overlays neutral mutations using coalescent methods and generates various summary statistics
- `Demography`: creates subpopulation demoography for SLiM based on phylogeny input by user
- `MutationType`: allows the user to create custom mutation types (*for exons and introns only*). Also allows the user to view the distribution of fitness effects based on mutation settings
- `MutationList`: user-defined list of all the mutations they would like to include in their simulation
- `ElementType`: allows the user to create custom genomic element types using the custom mutation types generated with `MutationType` class
- `ElementList`: user-defined list of all the element types they want included in their chromosome
- `Build`: allows the user to generate a chromosome based on their specifications; incorporates custom element types provided by an `ElementList` object
- `Chromosome`: reads in Build class objects and allows the user to view the chromosome before using it in the simulation


**Parameters** (User-controlled):
* `lifehistory`:selection will act differently on haploid and diploid life stages depending on the kind of lineage the user wants to model
* `tree`: phylogeny (newick or .tree object from toytree) that is used to generate subpopulation demography
* `genome_size`: the size of the haploid set of genetic material upon which evolution is being simulated
* `recomb`: recombination rate
*`generations`: length of the simulation in generations (if no tree is provided)
* `Ne`: effective population size (under the hood, acts similarly to a 'carrying capacity')
* `chromosome`: allows user to make adjustments to linear arrangement of genomes into genes and other regions
	- `elementlist`: list of genomic elements types that can be used to build the chromosome
		- `mutationlist`: types of mutations that can occur in genomic element types in the simulation
		- `mutationrate`: defined as MutationTypes are created


**Parameters** (not editable by user):
- `model`: SLiM3 can use Wright-Fisher (WF), extended WF, or non Wright-Fisher models. `shadie` will always run a nonWF model
- `reproduction`: pre-defined mating models, e.g. clonal, slefing, hermaphroditic, sexual, and combinations thereof
- `dispatch`: the dispatch script generated by `Script` class; may include instructions for collecting summary statistics too

~

### Description of the data:
Users will be able to input a phylogeny and the program will generate the corresponding population demography for input into SLiM. 

The `Demography` submodule of `shadie` will take the phylogeny from the user (provided as newick string or toytree object). It will use `toytree` to traverse the tree from root to tips and collect information it needs to generate the SLiM demography in a `pandas` dataframe, for example:

| source    | child2    | nodeheight| gen       |
| --------- | --------- | --------- | --------- |
| 0         | 3         | 1000000   | 1         |
| 3         | 5         | 600000    | 801       |
| 0         | 2         | 600000    | 801       |
| 3         | 4         | 300000    | 1401      |
| 0         | 1         | 300000    | 1401      |
 

- each node moving from tips to root is renamed with the lowest number of the child tips
- `gen` = 1+(abs(nodeheight-treeheight)*(gentime/treeheight))
- This example data is from a tree generated in toytree, which has no outgroup. Because the first split happens at gen1, we probably want a burn-in time before the simulation begins. 

 This will be used to generate this part of the script, which controls when populations in the program split into subpopulations:

Beyond that, `shadie` will largely generate its own data and output summary statistics that can be used to evaluate observed data. The user will choose from a relatively simple set of options.

~

### Demonstration of user interaction
The user can optionally provide a phylogeny, which will control the demography of the simulation (detailed above).

User can then adjust other simulation parameters, such as life history, mutation rate, and recombination rate. `package` has built-in defaults for many values (which are adjusted dynamically by the script based on the parameters the user *does* define), so the user does not have to define every parameter for every run. 

Example of full `shadie` workflow:

1. User defines mutation types and mutation list
```python
from shadie import MutationType
from shadie import Mutation List



```

2. User defines element types and element list
```python
from shadie import ElementType
from shadie import ElementList



```

3. User provides the chromosome structure they would like
```python
from shadie import Build



```
	Note: this is just one example of how to create a chromosome in `shadie`. Refer to notebook #3 in `notebooks` directory for a full demonstration of all the features of the `Build` class. 

4. User initializes a `Chromosome` class object using the `Build` object specified in step #3. They can use the functions of this class to review all the settings of their genome before proceeding

```python
from shadie import Chromosome

mychromosome
```

5. User converts their phylogeny to `toytree` object 
```python
import toytree

#option 1: read in existing newick file

my tree = 

#option 2: create a random tree

mytree = 

```

6. User sets up the script
```python
from shadie import Shadie

sim1 = Shadie(
	tree =  mytree				#must be toytree object
	chromosome = mychromosome	#must be Chromosome object
	reproduction = "pter",
	recomb=1e-9,
	genome_size=1e6)

#the write() function generates the slim script:
sim1.write()
```

7. Once the user is finished preparing the simulation script they can run a simulation in SLiM3:
```python
sim1.run()
```

8. When the simulation is finished, the user can initialize a `PostSim` class object, which will overlay mutations using `msprime`

```python
from shadie import PostSim

sim_out = PostSim(sim1)
```

```python
sim_out.summary()
```


*Coming soon*: user can calculate dN/dS, perform an MK-test, visualize mutations on a phylogeny, etc...
```python
# example of summary statistics from SLiM output
#MK-test
sim1_out.mk

#dN/dS
sim1_out.dnds

#draw gene trees?
sim1out.draw
...
```
~

### Similar Tools
[`pyslim`](https://github.com/tskit-dev/pyslim) is a python module that handles tree sequence files output by SLiM as a thin interface with `tskit`. This module will make it much easier to generate summary statistics for the user, as they've already done a lot of the work. `pyslim` will also allow `shadie` to use a combination of SLiM forward-in-time simulation and coalescent in `msprime` to run simulations more quickly. 

[SLiMgui](https://messerlab.org/slim/) is provided with SLiM3 installation on MacOSX. The SLiMgui provides a number of nice visualizations, some of which are similar to those in `shadie`. However, the SLiMgui still requires the user to hand-code the .slim script whereas `shadie` automates the script-writing process. 

To my knowledge, there are no other publicly available python wrappers for SLiM3 and no other evolution simulation programs that were built to model selection on biphasic lifecycles. 
