{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Targeted Differential Expression with `shadie`\n",
    "This notebook re-creates the simulations used in [manuscript]. There are only two genes on the chromosome, both of which only experience beneficial mutations. One is only expressed during the haploid life stage and the other is only expressed during the diploid life stage. \n",
    "This simulation takes advantage of advanced features of shadie, such as differential expression "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import shadie"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make Custom Mutations\n",
    "First, we create two mutation types with differential expression - one mutation type will only affect the haploid life stage and the other will only affect the diploid life stage. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "m1 = shadie.mtype(0.5, 'f', 0.1, diffexpr=\"haploid\") #this mutation type only expressed in haploids\n",
    "m2 = shadie.mtype(0.5, 'f', 0.1, diffexpr=\"diploid\") #this mutation type only expressed in diploids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make Custom Genomic Elements\n",
    "Here, we create two small genomic elements. One will only be expressed duriing the haploid life stage and one during the diploid life stage - this functionality is a result of the mutation types used when defining the genomic element. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "e1 = shadie.etype([m1], [1]) #inherits m1 __; only expressed in haploids\n",
    "e2 = shadie.etype([m2], [1]) #inherits m2 ___; only expressed in diploids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build the Chromosome\n",
    "Finally, we use our genomic elements to define genes on our chromosome. Each is 1000 base pairs long. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom = shadie.chromosome.explicit(data = {(0, 5000000): shadie.NONCDS,\n",
    "                                           (5000001, 5001000):e1,\n",
    "                                           (5001001, 10000000): shadie.NONCDS,\n",
    "                                           (10000001, 10001000):e2,\n",
    "                                           (10001001, 15000000): shadie.NONCDS},\n",
    "                                   use_synonymous_sites_in_coding=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(<toyplot.canvas.Canvas at 0x7fb40ac11250>,\n",
       " <toyplot.coordinates.Cartesian at 0x7fb40ac112e0>,\n",
       " <toyplot.mark.FillBoundaries at 0x7fb40ac5f2e0>)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div class=\"toyplot\" id=\"t4ebf92b9b84d446dad56fc219e251bf4\" style=\"text-align:center\"><svg class=\"toyplot-canvas-Canvas\" xmlns:toyplot=\"http://www.sandia.gov/toyplot\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" xmlns=\"http://www.w3.org/2000/svg\" width=\"700.0px\" height=\"150.0px\" viewBox=\"0 0 700.0 150.0\" preserveAspectRatio=\"xMidYMid meet\" style=\"background-color:transparent;border-color:#292724;border-style:none;border-width:1.0;fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:Helvetica;font-size:12px;opacity:1.0;stroke:rgb(16.1%,15.3%,14.1%);stroke-opacity:1.0;stroke-width:1.0\" id=\"t7d6da3538f2047d286dc53bb6f1b79cc\"><g class=\"toyplot-coordinates-Cartesian\" id=\"tf46a2d81d5304a58bce171bf8fe2e0eb\"><clipPath id=\"t5d4bbff96aea4f8e961fa6bc87e1dc17\"><rect x=\"40.0\" y=\"40.0\" width=\"620.0\" height=\"70.0\"></rect></clipPath><g clip-path=\"url(#t5d4bbff96aea4f8e961fa6bc87e1dc17)\"><g class=\"toyplot-mark-FillBoundaries\" style=\"stroke:rgb(0%,0%,0%);stroke-opacity:1.0\" id=\"tff7b8fdc60af482090171c30c0663f76\"><polygon points=\"50.0,100.0 250.0,100.0 250.0,50.0 50.0,50.0\" style=\"fill:rgb(16.9%,51.4%,72.9%);fill-opacity:1.0;opacity:0.75;stroke:rgb(0%,0%,0%);stroke-opacity:1.0\"><title>name: noncds\n",
       "interval: (0, 5000000)\n",
       "ElementType: g3\n",
       "coding: False</title></polygon></g><g class=\"toyplot-mark-FillBoundaries\" style=\"stroke:rgb(0%,0%,0%);stroke-opacity:1.0\" id=\"t2e23e3ed8a30411c87544fa654e6e372\"><polygon points=\"250.00003999999998,100.0 250.04,100.0 250.04,50.0 250.00003999999998,50.0\" style=\"fill:rgb(67.1%,86.7%,64.3%);fill-opacity:1.0;opacity:0.75;stroke:rgb(0%,0%,0%);stroke-opacity:1.0\"><title>name: None\n",
       "interval: (5000001, 5001000)\n",
       "ElementType: g5\n",
       "coding: True</title></polygon></g><g class=\"toyplot-mark-FillBoundaries\" style=\"stroke:rgb(0%,0%,0%);stroke-opacity:1.0\" id=\"td2ae93c79e0d4b0ca620869e05c72fed\"><polygon points=\"250.04003999999998,100.0 450.0,100.0 450.0,50.0 250.04003999999998,50.0\" style=\"fill:rgb(16.9%,51.4%,72.9%);fill-opacity:1.0;opacity:0.75;stroke:rgb(0%,0%,0%);stroke-opacity:1.0\"><title>name: noncds\n",
       "interval: (5001001, 10000000)\n",
       "ElementType: g3\n",
       "coding: False</title></polygon></g><g class=\"toyplot-mark-FillBoundaries\" style=\"stroke:rgb(0%,0%,0%);stroke-opacity:1.0\" id=\"t913e50d716c34f84951f8159c7b62ddc\"><polygon points=\"450.00004,100.0 450.04,100.0 450.04,50.0 450.00004,50.0\" style=\"fill:rgb(99.2%,68.2%,38%);fill-opacity:1.0;opacity:0.75;stroke:rgb(0%,0%,0%);stroke-opacity:1.0\"><title>name: None\n",
       "interval: (10000001, 10001000)\n",
       "ElementType: g6\n",
       "coding: True</title></polygon></g><g class=\"toyplot-mark-FillBoundaries\" style=\"stroke:rgb(0%,0%,0%);stroke-opacity:1.0\" id=\"t232b376e187144728117a8076386b551\"><polygon points=\"450.04004,100.0 650.0,100.0 650.0,50.0 450.04004,50.0\" style=\"fill:rgb(16.9%,51.4%,72.9%);fill-opacity:1.0;opacity:0.75;stroke:rgb(0%,0%,0%);stroke-opacity:1.0\"><title>name: noncds\n",
       "interval: (10001001, 15000000)\n",
       "ElementType: g3\n",
       "coding: False</title></polygon></g></g><g class=\"toyplot-coordinates-Axis\" id=\"tcf2471347f974c459b58aed75264fa00\" transform=\"translate(50.0,100.0)translate(0,10.0)\"><line x1=\"0\" y1=\"0\" x2=\"600.0\" y2=\"0\" style=\"\"></line><g><line x1=\"0.0\" y1=\"0\" x2=\"0.0\" y2=\"-5\" style=\"\"></line><line x1=\"100.0\" y1=\"0\" x2=\"100.0\" y2=\"-5\" style=\"\"></line><line x1=\"200.0\" y1=\"0\" x2=\"200.0\" y2=\"-5\" style=\"\"></line><line x1=\"300.0\" y1=\"0\" x2=\"300.0\" y2=\"-5\" style=\"\"></line><line x1=\"400.0\" y1=\"0\" x2=\"400.0\" y2=\"-5\" style=\"\"></line><line x1=\"500.0\" y1=\"0\" x2=\"500.0\" y2=\"-5\" style=\"\"></line><line x1=\"600.0\" y1=\"0\" x2=\"600.0\" y2=\"-5\" style=\"\"></line></g><g><g transform=\"translate(0.0,6)\"><text x=\"-2.78\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">0</text></g><g transform=\"translate(100.0,6)\"><text x=\"-19.459999999999997\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">2500000</text></g><g transform=\"translate(200.0,6)\"><text x=\"-19.459999999999997\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">5000000</text></g><g transform=\"translate(300.0,6)\"><text x=\"-19.459999999999997\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">7500000</text></g><g transform=\"translate(400.0,6)\"><text x=\"-22.24\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">10000000</text></g><g transform=\"translate(500.0,6)\"><text x=\"-22.24\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">12500000</text></g><g transform=\"translate(600.0,6)\"><text x=\"-22.24\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">15000000</text></g></g><g class=\"toyplot-coordinates-Axis-coordinates\" style=\"visibility:hidden\" transform=\"\"><line x1=\"0\" x2=\"0\" y1=\"-3.0\" y2=\"4.5\" style=\"stroke:rgb(43.9%,50.2%,56.5%);stroke-opacity:1.0;stroke-width:1.0\"></line><text x=\"0\" y=\"-6\" style=\"alignment-baseline:alphabetic;fill:rgb(43.9%,50.2%,56.5%);fill-opacity:1.0;font-size:10px;font-weight:normal;stroke:none;text-anchor:middle\"></text></g></g></g></svg><div class=\"toyplot-behavior\"><script>(function()\n",
       "{\n",
       "var modules={};\n",
       "modules[\"toyplot/tables\"] = (function()\n",
       "    {\n",
       "        var tables = [];\n",
       "\n",
       "        var module = {};\n",
       "\n",
       "        module.set = function(owner, key, names, columns)\n",
       "        {\n",
       "            tables.push({owner: owner, key: key, names: names, columns: columns});\n",
       "        }\n",
       "\n",
       "        module.get = function(owner, key)\n",
       "        {\n",
       "            for(var i = 0; i != tables.length; ++i)\n",
       "            {\n",
       "                var table = tables[i];\n",
       "                if(table.owner != owner)\n",
       "                    continue;\n",
       "                if(table.key != key)\n",
       "                    continue;\n",
       "                return {names: table.names, columns: table.columns};\n",
       "            }\n",
       "        }\n",
       "\n",
       "        module.get_csv = function(owner, key)\n",
       "        {\n",
       "            var table = module.get(owner, key);\n",
       "            if(table != undefined)\n",
       "            {\n",
       "                var csv = \"\";\n",
       "                csv += table.names.join(\",\") + \"\\n\";\n",
       "                for(var i = 0; i != table.columns[0].length; ++i)\n",
       "                {\n",
       "                  for(var j = 0; j != table.columns.length; ++j)\n",
       "                  {\n",
       "                    if(j)\n",
       "                      csv += \",\";\n",
       "                    csv += table.columns[j][i];\n",
       "                  }\n",
       "                  csv += \"\\n\";\n",
       "                }\n",
       "                return csv;\n",
       "            }\n",
       "        }\n",
       "\n",
       "        return module;\n",
       "    })();\n",
       "modules[\"toyplot/root/id\"] = \"t4ebf92b9b84d446dad56fc219e251bf4\";\n",
       "modules[\"toyplot/root\"] = (function(root_id)\n",
       "    {\n",
       "        return document.querySelector(\"#\" + root_id);\n",
       "    })(modules[\"toyplot/root/id\"]);\n",
       "modules[\"toyplot/canvas/id\"] = \"t7d6da3538f2047d286dc53bb6f1b79cc\";\n",
       "modules[\"toyplot/canvas\"] = (function(canvas_id)\n",
       "    {\n",
       "        return document.querySelector(\"#\" + canvas_id);\n",
       "    })(modules[\"toyplot/canvas/id\"]);\n",
       "modules[\"toyplot/menus/context\"] = (function(root, canvas)\n",
       "    {\n",
       "        var wrapper = document.createElement(\"div\");\n",
       "        wrapper.innerHTML = \"<ul class='toyplot-context-menu' style='background:#eee; border:1px solid #b8b8b8; border-radius:5px; box-shadow: 0px 0px 8px rgba(0%,0%,0%,0.25); margin:0; padding:3px 0; position:fixed; visibility:hidden;'></ul>\"\n",
       "        var menu = wrapper.firstChild;\n",
       "\n",
       "        root.appendChild(menu);\n",
       "\n",
       "        var items = [];\n",
       "\n",
       "        var ignore_mouseup = null;\n",
       "        function open_menu(e)\n",
       "        {\n",
       "            var show_menu = false;\n",
       "            for(var index=0; index != items.length; ++index)\n",
       "            {\n",
       "                var item = items[index];\n",
       "                if(item.show(e))\n",
       "                {\n",
       "                    item.item.style.display = \"block\";\n",
       "                    show_menu = true;\n",
       "                }\n",
       "                else\n",
       "                {\n",
       "                    item.item.style.display = \"none\";\n",
       "                }\n",
       "            }\n",
       "\n",
       "            if(show_menu)\n",
       "            {\n",
       "                ignore_mouseup = true;\n",
       "                menu.style.left = (e.clientX + 1) + \"px\";\n",
       "                menu.style.top = (e.clientY - 5) + \"px\";\n",
       "                menu.style.visibility = \"visible\";\n",
       "                e.stopPropagation();\n",
       "                e.preventDefault();\n",
       "            }\n",
       "        }\n",
       "\n",
       "        function close_menu()\n",
       "        {\n",
       "            menu.style.visibility = \"hidden\";\n",
       "        }\n",
       "\n",
       "        function contextmenu(e)\n",
       "        {\n",
       "            open_menu(e);\n",
       "        }\n",
       "\n",
       "        function mousemove(e)\n",
       "        {\n",
       "            ignore_mouseup = false;\n",
       "        }\n",
       "\n",
       "        function mouseup(e)\n",
       "        {\n",
       "            if(ignore_mouseup)\n",
       "            {\n",
       "                ignore_mouseup = false;\n",
       "                return;\n",
       "            }\n",
       "            close_menu();\n",
       "        }\n",
       "\n",
       "        function keydown(e)\n",
       "        {\n",
       "            if(e.key == \"Escape\" || e.key == \"Esc\" || e.keyCode == 27)\n",
       "            {\n",
       "                close_menu();\n",
       "            }\n",
       "        }\n",
       "\n",
       "        canvas.addEventListener(\"contextmenu\", contextmenu);\n",
       "        canvas.addEventListener(\"mousemove\", mousemove);\n",
       "        document.addEventListener(\"mouseup\", mouseup);\n",
       "        document.addEventListener(\"keydown\", keydown);\n",
       "\n",
       "        var module = {};\n",
       "        module.add_item = function(label, show, activate)\n",
       "        {\n",
       "            var wrapper = document.createElement(\"div\");\n",
       "            wrapper.innerHTML = \"<li class='toyplot-context-menu-item' style='background:#eee; color:#333; padding:2px 20px; list-style:none; margin:0; text-align:left;'>\" + label + \"</li>\"\n",
       "            var item = wrapper.firstChild;\n",
       "\n",
       "            items.push({item: item, show: show});\n",
       "\n",
       "            function mouseover()\n",
       "            {\n",
       "                this.style.background = \"steelblue\";\n",
       "                this.style.color = \"white\";\n",
       "            }\n",
       "\n",
       "            function mouseout()\n",
       "            {\n",
       "                this.style.background = \"#eee\";\n",
       "                this.style.color = \"#333\";\n",
       "            }\n",
       "\n",
       "            function choose_item(e)\n",
       "            {\n",
       "                close_menu();\n",
       "                activate();\n",
       "\n",
       "                e.stopPropagation();\n",
       "                e.preventDefault();\n",
       "            }\n",
       "\n",
       "            item.addEventListener(\"mouseover\", mouseover);\n",
       "            item.addEventListener(\"mouseout\", mouseout);\n",
       "            item.addEventListener(\"mouseup\", choose_item);\n",
       "            item.addEventListener(\"contextmenu\", choose_item);\n",
       "\n",
       "            menu.appendChild(item);\n",
       "        };\n",
       "        return module;\n",
       "    })(modules[\"toyplot/root\"],modules[\"toyplot/canvas\"]);\n",
       "modules[\"toyplot/io\"] = (function()\n",
       "    {\n",
       "        var module = {};\n",
       "        module.save_file = function(mime_type, charset, data, filename)\n",
       "        {\n",
       "            var uri = \"data:\" + mime_type + \";charset=\" + charset + \",\" + data;\n",
       "            uri = encodeURI(uri);\n",
       "\n",
       "            var link = document.createElement(\"a\");\n",
       "            if(typeof link.download != \"undefined\")\n",
       "            {\n",
       "              link.href = uri;\n",
       "              link.style = \"visibility:hidden\";\n",
       "              link.download = filename;\n",
       "\n",
       "              document.body.appendChild(link);\n",
       "              link.click();\n",
       "              document.body.removeChild(link);\n",
       "            }\n",
       "            else\n",
       "            {\n",
       "              window.open(uri);\n",
       "            }\n",
       "        };\n",
       "        return module;\n",
       "    })();\n",
       "modules[\"toyplot.coordinates.Axis\"] = (\n",
       "        function(canvas)\n",
       "        {\n",
       "            function sign(x)\n",
       "            {\n",
       "                return x < 0 ? -1 : x > 0 ? 1 : 0;\n",
       "            }\n",
       "\n",
       "            function mix(a, b, amount)\n",
       "            {\n",
       "                return ((1.0 - amount) * a) + (amount * b);\n",
       "            }\n",
       "\n",
       "            function log(x, base)\n",
       "            {\n",
       "                return Math.log(Math.abs(x)) / Math.log(base);\n",
       "            }\n",
       "\n",
       "            function in_range(a, x, b)\n",
       "            {\n",
       "                var left = Math.min(a, b);\n",
       "                var right = Math.max(a, b);\n",
       "                return left <= x && x <= right;\n",
       "            }\n",
       "\n",
       "            function inside(range, projection)\n",
       "            {\n",
       "                for(var i = 0; i != projection.length; ++i)\n",
       "                {\n",
       "                    var segment = projection[i];\n",
       "                    if(in_range(segment.range.min, range, segment.range.max))\n",
       "                        return true;\n",
       "                }\n",
       "                return false;\n",
       "            }\n",
       "\n",
       "            function to_domain(range, projection)\n",
       "            {\n",
       "                for(var i = 0; i != projection.length; ++i)\n",
       "                {\n",
       "                    var segment = projection[i];\n",
       "                    if(in_range(segment.range.bounds.min, range, segment.range.bounds.max))\n",
       "                    {\n",
       "                        if(segment.scale == \"linear\")\n",
       "                        {\n",
       "                            var amount = (range - segment.range.min) / (segment.range.max - segment.range.min);\n",
       "                            return mix(segment.domain.min, segment.domain.max, amount)\n",
       "                        }\n",
       "                        else if(segment.scale[0] == \"log\")\n",
       "                        {\n",
       "                            var amount = (range - segment.range.min) / (segment.range.max - segment.range.min);\n",
       "                            var base = segment.scale[1];\n",
       "                            return sign(segment.domain.min) * Math.pow(base, mix(log(segment.domain.min, base), log(segment.domain.max, base), amount));\n",
       "                        }\n",
       "                    }\n",
       "                }\n",
       "            }\n",
       "\n",
       "            var axes = {};\n",
       "\n",
       "            function display_coordinates(e)\n",
       "            {\n",
       "                var current = canvas.createSVGPoint();\n",
       "                current.x = e.clientX;\n",
       "                current.y = e.clientY;\n",
       "\n",
       "                for(var axis_id in axes)\n",
       "                {\n",
       "                    var axis = document.querySelector(\"#\" + axis_id);\n",
       "                    var coordinates = axis.querySelector(\".toyplot-coordinates-Axis-coordinates\");\n",
       "                    if(coordinates)\n",
       "                    {\n",
       "                        var projection = axes[axis_id];\n",
       "                        var local = current.matrixTransform(axis.getScreenCTM().inverse());\n",
       "                        if(inside(local.x, projection))\n",
       "                        {\n",
       "                            var domain = to_domain(local.x, projection);\n",
       "                            coordinates.style.visibility = \"visible\";\n",
       "                            coordinates.setAttribute(\"transform\", \"translate(\" + local.x + \")\");\n",
       "                            var text = coordinates.querySelector(\"text\");\n",
       "                            text.textContent = domain.toFixed(2);\n",
       "                        }\n",
       "                        else\n",
       "                        {\n",
       "                            coordinates.style.visibility= \"hidden\";\n",
       "                        }\n",
       "                    }\n",
       "                }\n",
       "            }\n",
       "\n",
       "            canvas.addEventListener(\"click\", display_coordinates);\n",
       "\n",
       "            var module = {};\n",
       "            module.show_coordinates = function(axis_id, projection)\n",
       "            {\n",
       "                axes[axis_id] = projection;\n",
       "            }\n",
       "\n",
       "            return module;\n",
       "        })(modules[\"toyplot/canvas\"]);\n",
       "(function(tables, context_menu, io, owner_id, key, label, names, columns, filename)\n",
       "        {\n",
       "            tables.set(owner_id, key, names, columns);\n",
       "\n",
       "            var owner = document.querySelector(\"#\" + owner_id);\n",
       "            function show_item(e)\n",
       "            {\n",
       "                return owner.contains(e.target);\n",
       "            }\n",
       "\n",
       "            function choose_item()\n",
       "            {\n",
       "                io.save_file(\"text/csv\", \"utf-8\", tables.get_csv(owner_id, key), filename + \".csv\");\n",
       "            }\n",
       "\n",
       "            context_menu.add_item(\"Save \" + label + \" as CSV\", show_item, choose_item);\n",
       "        })(modules[\"toyplot/tables\"],modules[\"toyplot/menus/context\"],modules[\"toyplot/io\"],\"tff7b8fdc60af482090171c30c0663f76\",\"data\",\"fill data\",[\"x\", \"y0\", \"y1\"],[[0.0, 5000000.0], [0.0, 0.0], [1.0, 1.0]],\"toyplot\");\n",
       "(function(tables, context_menu, io, owner_id, key, label, names, columns, filename)\n",
       "        {\n",
       "            tables.set(owner_id, key, names, columns);\n",
       "\n",
       "            var owner = document.querySelector(\"#\" + owner_id);\n",
       "            function show_item(e)\n",
       "            {\n",
       "                return owner.contains(e.target);\n",
       "            }\n",
       "\n",
       "            function choose_item()\n",
       "            {\n",
       "                io.save_file(\"text/csv\", \"utf-8\", tables.get_csv(owner_id, key), filename + \".csv\");\n",
       "            }\n",
       "\n",
       "            context_menu.add_item(\"Save \" + label + \" as CSV\", show_item, choose_item);\n",
       "        })(modules[\"toyplot/tables\"],modules[\"toyplot/menus/context\"],modules[\"toyplot/io\"],\"t2e23e3ed8a30411c87544fa654e6e372\",\"data\",\"fill data\",[\"x\", \"y0\", \"y1\"],[[5000001.0, 5001000.0], [0.0, 0.0], [1.0, 1.0]],\"toyplot\");\n",
       "(function(tables, context_menu, io, owner_id, key, label, names, columns, filename)\n",
       "        {\n",
       "            tables.set(owner_id, key, names, columns);\n",
       "\n",
       "            var owner = document.querySelector(\"#\" + owner_id);\n",
       "            function show_item(e)\n",
       "            {\n",
       "                return owner.contains(e.target);\n",
       "            }\n",
       "\n",
       "            function choose_item()\n",
       "            {\n",
       "                io.save_file(\"text/csv\", \"utf-8\", tables.get_csv(owner_id, key), filename + \".csv\");\n",
       "            }\n",
       "\n",
       "            context_menu.add_item(\"Save \" + label + \" as CSV\", show_item, choose_item);\n",
       "        })(modules[\"toyplot/tables\"],modules[\"toyplot/menus/context\"],modules[\"toyplot/io\"],\"td2ae93c79e0d4b0ca620869e05c72fed\",\"data\",\"fill data\",[\"x\", \"y0\", \"y1\"],[[5001001.0, 10000000.0], [0.0, 0.0], [1.0, 1.0]],\"toyplot\");\n",
       "(function(tables, context_menu, io, owner_id, key, label, names, columns, filename)\n",
       "        {\n",
       "            tables.set(owner_id, key, names, columns);\n",
       "\n",
       "            var owner = document.querySelector(\"#\" + owner_id);\n",
       "            function show_item(e)\n",
       "            {\n",
       "                return owner.contains(e.target);\n",
       "            }\n",
       "\n",
       "            function choose_item()\n",
       "            {\n",
       "                io.save_file(\"text/csv\", \"utf-8\", tables.get_csv(owner_id, key), filename + \".csv\");\n",
       "            }\n",
       "\n",
       "            context_menu.add_item(\"Save \" + label + \" as CSV\", show_item, choose_item);\n",
       "        })(modules[\"toyplot/tables\"],modules[\"toyplot/menus/context\"],modules[\"toyplot/io\"],\"t913e50d716c34f84951f8159c7b62ddc\",\"data\",\"fill data\",[\"x\", \"y0\", \"y1\"],[[10000001.0, 10001000.0], [0.0, 0.0], [1.0, 1.0]],\"toyplot\");\n",
       "(function(tables, context_menu, io, owner_id, key, label, names, columns, filename)\n",
       "        {\n",
       "            tables.set(owner_id, key, names, columns);\n",
       "\n",
       "            var owner = document.querySelector(\"#\" + owner_id);\n",
       "            function show_item(e)\n",
       "            {\n",
       "                return owner.contains(e.target);\n",
       "            }\n",
       "\n",
       "            function choose_item()\n",
       "            {\n",
       "                io.save_file(\"text/csv\", \"utf-8\", tables.get_csv(owner_id, key), filename + \".csv\");\n",
       "            }\n",
       "\n",
       "            context_menu.add_item(\"Save \" + label + \" as CSV\", show_item, choose_item);\n",
       "        })(modules[\"toyplot/tables\"],modules[\"toyplot/menus/context\"],modules[\"toyplot/io\"],\"t232b376e187144728117a8076386b551\",\"data\",\"fill data\",[\"x\", \"y0\", \"y1\"],[[10001001.0, 15000000.0], [0.0, 0.0], [1.0, 1.0]],\"toyplot\");\n",
       "(function(axis, axis_id, projection)\n",
       "        {\n",
       "            axis.show_coordinates(axis_id, projection);\n",
       "        })(modules[\"toyplot.coordinates.Axis\"],\"tcf2471347f974c459b58aed75264fa00\",[{\"domain\": {\"bounds\": {\"max\": Infinity, \"min\": -Infinity}, \"max\": 15000000.0, \"min\": 0.0}, \"range\": {\"bounds\": {\"max\": Infinity, \"min\": -Infinity}, \"max\": 600.0, \"min\": 0.0}, \"scale\": \"linear\"}]);\n",
       "})();</script></div></div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#when we visualiize our chromosome we can see two tiny genes at 5e6 and 10e6\n",
    "chrom.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create the Simulation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "initialize() {\n",
      "\n",
      "  // model type\n",
      "  initializeSLiMModelType(\"nonWF\");\n",
      "\n",
      "  // config\n",
      "  initializeRecombinationRate(1e-07, 15000000);\n",
      "  initializeMutationRate(1e-06);\n",
      "  initializeTreeSeq(simplificationInterval=200);\n",
      "\n",
      "  // MutationType init\n",
      "  initializeMutationType('m1', 0.5, 'f', 0.0);\n",
      "    m1.convertToSubstitution = T;\n",
      "  initializeMutationType('m4', 0.5, 'f', 0.1);\n",
      "    m4.convertToSubstitution = T;\n",
      "  initializeMutationType('m5', 0.5, 'f', 0.1);\n",
      "    m5.convertToSubstitution = T;\n",
      "  c(m1, m5, m4).haploidDominanceCoeff = 1.0;\n",
      "\n",
      "  // ElementType init\n",
      "  initializeGenomicElementType('g3', m1, 1);\n",
      "  initializeGenomicElementType('g5', m4, 1);\n",
      "  initializeGenomicElementType('g6', m5, 1);\n",
      "\n",
      "  // Chromosome (GenomicElement init)\n",
      "  initializeGenomicElement(g3, 0, 5000000);\n",
      "  initializeGenomicElement(g5, 5000001, 5001000);\n",
      "  initializeGenomicElement(g3, 5001001, 10000000);\n",
      "  initializeGenomicElement(g6, 10000001, 10001000);\n",
      "  initializeGenomicElement(g3, 10001001, 15000000);\n",
      "\n",
      "  // constants (Population sizes and others)\n",
      "  defineConstant('SPO_POP_SIZE', 3000);\n",
      "  defineConstant('GAM_POP_SIZE', 3000);\n",
      "  defineConstant('SPO_MUTATION_RATE', 5e-07);\n",
      "  defineConstant('GAM_MUTATION_RATE', 5e-07);\n",
      "  defineConstant('GAM_CLONE_RATE', 0.2);\n",
      "  defineConstant('GAM_CLONES_PER', 1);\n",
      "  defineConstant('GAM_SELF_RATE_PER_EGG', 0.0);\n",
      "  defineConstant('GAM_SELF_RATE', 0.2);\n",
      "  defineConstant('SPO_SELF_RATE_PER_EGG', 0.0);\n",
      "  defineConstant('SPO_SELF_RATE', 0.0);\n",
      "  defineConstant('SPO_RANDOM_DEATH_CHANCE', 0.0);\n",
      "  defineConstant('GAM_RANDOM_DEATH_CHANCE', 0.0);\n",
      "  defineConstant('SPO_SPORES_PER', 10);\n",
      "  defineConstant('GAM_MATERNAL_EFFECT', 0.0);\n",
      "  defineConstant('GAM_ARCHEGONIA_PER', 10);\n",
      "  defineConstant('GAM_K', 30000);\n",
      "\n",
      "  // extra scripts (Optional)\n",
      "  \n",
      "}\n",
      "\n",
      "// shadie DEFINITIONS\n",
      "// shadie DEFINITIONS\n",
      "// p0 = haploid population\n",
      "// p1 = diploid population\n",
      "// >1M- tmp sex tags\n",
      "// >2000000 = male gametophyte (1N)tag\n",
      "// <2000000 = female gametophyte (1N)tag\n",
      "// 2 = gametophyte clones (1N) tag\n",
      "// 3 = sporophyte (2N) tag\n",
      "// 4 = sporophyte clones (2N) tag -------------- None in this model;\n",
      "\n",
      "// generates offspring\n",
      "// generates sporophytes from gametes\n",
      "s5 reproduction(p0 ) {\n",
      "    // archegonia only produce one egg in bryophytes\n",
      "    eggs = GAM_ARCHEGONIA_PER;\n",
      "    \n",
      "    // clonal individual get added to the p0 pool for next round.\n",
      "    // NOTE: this doesn't allow clones to reproduce this round.\n",
      "    if (runif(1) < GAM_CLONE_RATE) {\n",
      "        for (i in 1:GAM_CLONES_PER) {\n",
      "            child = p0.addRecombinant(individual.genome1, NULL, NULL, NULL, NULL, NULL);\n",
      "            child.tag = 2;\n",
      "            child.setValue(\"parentid\", individual.tag);\n",
      "            if (GAM_MATERNAL_EFFECT > 0)\n",
      "                child.setValue(\"maternal_fitness\", NULL);\n",
      "        }\n",
      "    }\n",
      "    \n",
      "    // the original plant making the clones can still reproduce regardless \n",
      "    // of whether or not it cloned.\n",
      "    // iterate over each egg to find a mate (self, sib, or outcross)\n",
      "    // NOTE: each gametophyte gives rise to antheridia that produce thousands of\n",
      "    // clonal sperm. Because of this, sperm is not removed from the mating pool when used. \n",
      "    \n",
      "    //Reproduction scripts run only female gametophytes\n",
      "    if (individual.tag < 2000000) {\n",
      "        \n",
      "        // archegonia only produce one egg in bryophytes\n",
      "        eggs = GAM_ARCHEGONIA_PER;\n",
      "        \n",
      "        for (rep in 1:eggs) {\n",
      "            \n",
      "            // idea: random_chance_of ... here?\n",
      "            // weighted sampling: each egg gets gam-selfed, spo-selfed, or outcrossed.\n",
      "            // Note: shadie enforces that GAM_SELF_RATE + SPO_SELF_RATE !> 1\n",
      "            mode = sample(\n",
      "                x=c(1, 2, 3),\n",
      "                size=1,\n",
      "                weights=c(GAM_SELF_RATE_PER_EGG, SPO_SELF_RATE_PER_EGG, 1 - (GAM_SELF_RATE_PER_EGG + SPO_SELF_RATE_PER_EGG))\n",
      "                );\n",
      "            \n",
      "            // intra-gametic selfed (only possible in monoicous)\n",
      "            if (mode == 1) {\n",
      "                child = p1.addRecombinant(individual.genome1, NULL, NULL, individual.genome1, NULL, NULL);\n",
      "                child.tag = 3;\n",
      "                if (GAM_MATERNAL_EFFECT > 0)\n",
      "                    child.setValue(\"maternal_fitness\", subpop.cachedFitness(individual.index));\n",
      "            }\n",
      "            \n",
      "            // inter-gametophytic selfed individual (same sporo parent)\n",
      "            // only occurs IF a sibling gametophyte is still alive.\n",
      "            else if (mode == 2) {\n",
      "                sibling = p0.sampleIndividuals(1, tag=1000000+individual.tag);\n",
      "                if (sibling.size()>0) {\n",
      "                    child = p1.addRecombinant(individual.genome1, NULL, NULL, sibling.genome1, NULL, NULL);\n",
      "                    child.tag = 3;\n",
      "                    if (GAM_MATERNAL_EFFECT > 0)\n",
      "                        child.setValue(\"maternal_fitness\", subpop.cachedFitness(individual.index));\n",
      "                }\n",
      "            }\n",
      "            \n",
      "            // outcrossing individual samples any p0 that is not same tag.\n",
      "            // only occurs if a non-sib gametophyte is still alive.\n",
      "            else {\n",
      "                // try at most 10 times to find a non-sib sperm, then skip.\n",
      "                for (trial in 1:10) {\n",
      "                    sperm = sample(p0.individuals, 1);\n",
      "                    if (sperm.tag != individual.tag) {\n",
      "                        child = p1.addRecombinant(individual.genome1, NULL, NULL, sperm.genome1, NULL, NULL);\n",
      "                        child.tag = 3;\n",
      "                        if (GAM_MATERNAL_EFFECT > 0)\n",
      "                            child.setValue(\"maternal_fitness\", subpop.cachedFitness(individual.index));\n",
      "                        break;\n",
      "                    }\n",
      "                }\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "}\n",
      "\n",
      "\n",
      "// generates offspring\n",
      "// generates gametes from sporophytes\n",
      "s6 reproduction(p1 ) {\n",
      "    ind = individual;\n",
      "    \n",
      "    // parent tag is 1M + the parent's unique index\n",
      "    ptag = 1000000 + ind.index;\n",
      "    \n",
      "    // each spore produces its own recombinant breakpoints \n",
      "    for (rep in 1:SPO_SPORES_PER) {\n",
      "        breaks1 = sim.chromosome.drawBreakpoints(ind);\n",
      "        breaks2 = sim.chromosome.drawBreakpoints(ind);\n",
      "        \n",
      "        // create four meiotic products. If later two of these mate with each other\n",
      "        // it is an example of sporophytic selfing. Because we need to be able to match\n",
      "        // sibling gametes at that time we tag them now with their sporophyte parent's index.\n",
      "        p0.addRecombinant(ind.genome1, ind.genome2, breaks1, NULL, NULL, NULL).tag=ptag;\n",
      "        p0.addRecombinant(ind.genome2, ind.genome1, breaks1, NULL, NULL, NULL).tag=ptag;\n",
      "        p0.addRecombinant(ind.genome1, ind.genome2, breaks2, NULL, NULL, NULL).tag=ptag;\n",
      "        p0.addRecombinant(ind.genome2, ind.genome1, breaks2, NULL, NULL, NULL).tag=ptag;\n",
      "    }\n",
      "}\n",
      "\n",
      "\n",
      "// executes after offspring are generated\n",
      "// define subpops: p1=diploid sporophytes, p0=haploid gametophytes\n",
      "\n",
      "1 early() {\n",
      "    sim.addSubpop('p1', SPO_POP_SIZE);\n",
      "    sim.addSubpop('p0', 0);\n",
      "    p1.individuals.tag = 3;\n",
      "}\n",
      "\n",
      "\n",
      "// executes after offspring are generated\n",
      "// alternation of generations\n",
      "\n",
      "early() {\n",
      "    // diploids (p1) just generated haploid gametophytes\n",
      "    if (sim.generation % 2 == 0) {\n",
      "\n",
      "        // fitness affects gametophyte survival\n",
      "        p0.fitnessScaling = GAM_POP_SIZE / p0.individualCount;\n",
      "\n",
      "       //set mutation rate for haploids\n",
      "        sim.chromosome.setMutationRate(GAM_MUTATION_RATE);\n",
      "\n",
      "        // alternate generations.\n",
      "        // deactivate survival callback to remove inds from p1 subpop.\n",
      "        // activate survival callback to remove inds from p0 subpop.\n",
      "        s1.active = 1;\n",
      "        s2.active = 0;\n",
      "        \n",
      "        // add random chance of death, maternal effects, or fitness calculation.\n",
      "        // activate survival(p0) to set fitness method for p0s\n",
      "        // deactivate survival(p1) to set fitness method for p1s\n",
      "        s3.active = 1;\n",
      "        s4.active = 0;\n",
      "        \n",
      "        // set reproduction function to be used when next generation starts.\n",
      "        // deactivate reproduction(p1) diploids reproduce, haploids don't.\n",
      "        // activate reproduction(p0) haploids reproduce, diploids don't.        \n",
      "        s5.active = 1;\n",
      "        s6.active = 0;\n",
      "\n",
      "        s8.active = 1;\n",
      "        s7.active = 0;\n",
      "    }\n",
      "\n",
      "\n",
      "    // odd generations = gametophytes (p0) just generated sporophytes\n",
      "    else {\n",
      "\n",
      "        // fitness affects sporophytes\n",
      "        p1.fitnessScaling = SPO_POP_SIZE / p1.individualCount;\n",
      "        p0.fitnessScaling = GAM_K/ p0.individualCount;\n",
      "\n",
      "        // set mutation rate to sporophyte rate\n",
      "        sim.chromosome.setMutationRate(SPO_MUTATION_RATE);\n",
      "        \n",
      "        // alternate generations.\n",
      "        // deactivate survival(p0) to remove inds from p0 subpop (i.e., keep p0)\n",
      "        // activate survival(p1) to remove inds from p1 subpop.\n",
      "        s1.active = 0;\n",
      "        s2.active = 1;\n",
      "        \n",
      "        // add random chance of death, maternal effects, or fitness calculation.\n",
      "        // deactivate survival(p0) to set fitness method for p0s\n",
      "        // activate survival(p1) to set fitness method for p1s\n",
      "        s3.active = 0;\n",
      "        s4.active = 1;\n",
      "        \n",
      "        // set reproduction function to be used when next generation starts.\n",
      "        // activate reproduction(p1) diploids reproduce, haploids don't.\n",
      "        // deactivate reproduction(p0) haploids reproduce, diploids don't.        \n",
      "        s5.active = 0;\n",
      "        s6.active = 1;\n",
      "\n",
      "        s7.active = 1;\n",
      "        s8.active = 0;\n",
      "    }\n",
      "}\n",
      "\n",
      "// SURVIVAL CALLBACKS for alternation of generations\n",
      "// during haploid generation, remove p1 individuals unless it is a clone.\n",
      "s1 survival(p1) {\n",
      "    //clones survive to next gen and are re-tagged with parental tag\n",
      "    if (individual.tag == 4) {\n",
      "            individual.tag = individual.getValue(\"parentid\");\n",
      "            return T;\n",
      "        }\n",
      "    return F;\n",
      "}\n",
      "\n",
      "// alternate to other generation (diploid)\n",
      "// remove p0 individuals unless it is a clone.\n",
      "s2 survival(p0) {\n",
      "    // allow clones to persist until next generation, tag is reset to parent tag.\n",
      "    if (individual.tag == 2) {\n",
      "        individual.tag = individual.getValue(\"parentid\");\n",
      "        return T;\n",
      "    }\n",
      "    return F;\n",
      "}\n",
      "\n",
      "// remove p0s by three possible mechanisms: \n",
      "// 1. random chance of death.\n",
      "// 2. maternal effect.\n",
      "// 3. returns NULL, meaning use the standard fitness prob.\n",
      "s3 survival(p0) {\n",
      "    if (runif(1) < GAM_RANDOM_DEATH_CHANCE)\n",
      "        return F;\n",
      "    \n",
      "\t// Gametophyte mother fitness affects sporophyte survival\n",
      "    maternal_fitness = individual.getValue(\"maternal_fitness\");\n",
      "    if (!isNULL(maternal_fitness)) {\n",
      "        corrected_fitness = (maternal_fitness * GAM_MATERNAL_EFFECT) + fitness * (1 - GAM_MATERNAL_EFFECT);\n",
      "        return (runif(1) < corrected_fitness);\n",
      "    }\n",
      "\n",
      "    return NULL;\n",
      "}\n",
      "\n",
      "\n",
      "// remove p1s by three possible mechanisms: \n",
      "// 1. random chance of death.\n",
      "// 2. using fitness re-calculated to include maternal effect.\n",
      "// 3. returns NULL, meaning use the standard fitness prob.\n",
      "\n",
      "s4 survival(p1) {\n",
      "    if (runif(1) < SPO_RANDOM_DEATH_CHANCE)\n",
      "        return F;\n",
      "    \n",
      "\t// Gametophyte mother fitness affects sporophyte survival\n",
      "    maternal_fitness = individual.getValue(\"maternal_fitness\");\n",
      "    if (!isNULL(maternal_fitness)) {\n",
      "        corrected_fitness = (maternal_fitness * GAM_MATERNAL_EFFECT) + fitness * (1 - GAM_MATERNAL_EFFECT);\n",
      "        return (runif(1) < corrected_fitness);\n",
      "    }\n",
      "\n",
      "    return NULL;\n",
      "}\n",
      "\n",
      "// adjusts fitness calculation\n",
      "// turns expression off\n",
      "s7 fitness(m5) {\n",
      "    return 1.0;\n",
      "}\n",
      "\n",
      "\n",
      "// adjusts fitness calculation\n",
      "// turns expression off\n",
      "s8 fitness(m4) {\n",
      "    return 1.0;\n",
      "}\n",
      "\n",
      "\n",
      "// executes after selection occurs\n",
      "// end of sim; save .trees file\n",
      "\n",
      "10001 late() {\n",
      "    sim.treeSeqRememberIndividuals(sim.subpopulations.individuals);\n",
      "    sim.treeSeqOutput('target-bryo-mono.trees', \n",
      "metadata=Dictionary(\n",
      "\"spo_mutation_rate\", SPO_MUTATION_RATE,\n",
      "\"recombination_rate\", 1e-7,\n",
      "\"spo_population_size\", SPO_POP_SIZE,\n",
      "\"gam_population_size\", GAM_POP_SIZE)\n",
      ");\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with shadie.Model() as bryomono_model:\n",
    "    # init the model\n",
    "    bryomono_model.initialize(chromosome=chrom, \n",
    "                     sim_time = 10000, \n",
    "                     mutation_rate = 1e-6,\n",
    "                     recomb_rate = 1e-07,\n",
    "                     file_out = \"target-bryo-mono.trees\")\n",
    "\n",
    "    bryomono_model.reproduction.bryophyte_monoicous(\n",
    "        spo_pop_size=3000, \n",
    "        gam_pop_size=3000,\n",
    "        gam_clone_rate=0.2,\n",
    "        gam_clones_per=1,\n",
    "        spo_spores_per=10,\n",
    "        gam_archegonia_per=10,\n",
    "        \n",
    "    )\n",
    "#Uncomment line below to view the script\n",
    "#print(bryomono_model.script)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-d4ecfbdab951>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mbryomono_model\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/code/git/hacks/shadie/shadie/sims/model.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self, seed, binary)\u001b[0m\n\u001b[1;32m    462\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    463\u001b[0m             \u001b[0;31m# capture stdout\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 464\u001b[0;31m             \u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mproc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommunicate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    465\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    466\u001b[0m             \u001b[0;31m# check for errors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.8/subprocess.py\u001b[0m in \u001b[0;36mcommunicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1009\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_stdin_write\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1010\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1011\u001b[0;31m                 \u001b[0mstdout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1012\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1013\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "bryomono_model.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using the `PostSim` Module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyslim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bryomono = \"/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "post1 = shadie.postsim.OneSim(\n",
    "    bryomono,\n",
    "    chromosome=chrom, \n",
    "    mut=5e-7,\n",
    "    recomb=1e-7, \n",
    "    recapitate=True, \n",
    "    ancestral_Ne=3000, \n",
    "    add_neutral_mutations=True,\n",
    ")\n",
    "\n",
    "post1.tree_sequence\n",
    "#post.draw_tree_sequence(max_trees=10);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
